### ISSUE TITLE
davici API to unload single certificate from strongswan

### ISSUE THREAD

**Issue by amdrsantos** [strongswan/strongswan#928]
Currently, there is no davici API to remove just a single certificate from strongSwan.
The only way is to clear all credentials (clear-creds command) which unloads all certificates and private keys from strongswan, and then load the ones you actually want.
[Refer to How to unload a particular certificate from strongswan.](https://wiki.strongswan.org/issues/3680)

This is a limitation because when a certificate is unloaded, all traffic must be re-started, even if the certificate is not being used at all.

Is it possible to improve davici library to get rid of this limitation?

Thanks in advance,
Alex

**Comment by tobiasbrunner** [comment]
> This is a limitation because when a certificate is unloaded, all traffic must be re-started, even if the certificate is not being used at all.

What do you mean?

CA certificates can be unloaded individually since 5.9.0 (#172) via `load|unload-authority` (unless they are also loaded via `load-certs`). However, certificates loaded via `load-certs` or configs (which end up in the same credential set) currently can't be unloaded individually. It's not something that's usually necessary, so what exactly is your use case?

**Comment by amdrsantos** [comment]
Thanks for your clarification regarding the CA certificates.

But what about the leaf certificates?
Imagine that I have 2 connections to 2 different peers.
leaf certificate A is used to authenticate peer X
leaf certificate B is used to authenticate peer Y

Now, if I want to change the leaf certificate to authenticate peer X (for security reasons, or because the service in peer X changed server), I will have to do a clear-creds which removes all certificates, impacting traffic in peer Y, instead of just impacting traffic in peer X.

Thanks and Regards,
Alex

**Comment by tobiasbrunner** [comment]
> Now, if I want to change the leaf certificate to authenticate peer X (...), I will have to do a clear-creds which removes all certificates, impacting traffic in peer Y, instead of just impacting traffic in peer X.

No, that has no impact on the existing connection with peer X let alone its traffic. If you loaded the certificate with the connection (`remote<suffix>.certs`) it won't even have an impact on new connections with peer X. If you didn't, there could be a short window (until the certificate is loaded again) in which the certificate is not available and new connections with peer X would be affected (unless the peer does send its own certificate and it can be verified via CA certs).

`clear-creds` clears the credential set in the _vici_ plugin, which stores certificates loaded with `load-cert` and caches those loaded with `load-conn`. But those connections have their own references to these certificates, which are not affected by the command. If you modify such a connection later and reload it, there will be a new instance of the certificate, so that wouldn't be 100% ideal in regards to memory usage, but it does not affect the functionality.

> for security reasons,

You might want to look into OCSP and CRLs (note that these mechanisms are ignored if end-entity certificates as loaded locally, because those are always trusted ultimately, they could even be expired).

> because the service in peer X changed server

I assume you mean this changes the server's identity. In this case the already loaded certificate shouldn't be an issue because if the identity in the connection changes, the certificate won't match anymore.

**Comment by amdrsantos** [comment]
We tried the following vici commands:
1. load-key to load private key of the local certificate;
2. load-conn with connections.<conn>.local.cert<suffix>.file
Connection is established. 
And both certificates are in the cache.

Now we want to remove everything, by giving the commands:
3. unload-key()
4. terminate()
5. unload-conn()

Both certificates (local and remote) are still in cache.

6. flush-cert()
Even after flush-certs (trying to flush all certificates), the remote certificate is flushed, but the local certificate still in the cache.

So only with clear-creads(), it is possible to unload the certificate.

**Comment by Thermi** [comment]
Yeah so currently it's impossible to unload a single leaf certificate. Would that be possible to implement, Tobias?

**Comment by tobiasbrunner** [comment]
Technically, sure. But it's tricky as certificates can get loaded in different ways (via configs, directly via load-cert()) but are not explicitly tracked once they are added to the in-memory credential set (i.e. we don't know how many configs reference a certificate or whether it was also or only loaded manually). So when unloading a config or even manually unloading a certificate, it's not ideal to just remove it as other configs might reference it, which is fine for loaded configs as their reference to the existing certificate won't be invalidated. However, if a config is loaded later that uses the same certificate that will create a new instance of the certificate as any other won't be found in the credential set. Do that multiple times and you have kind of a memory leak. So this would require some refactoring in regards to how certificates are managed in the vici plugin. But since having to remove a certificate without a restart (or where flushing the creds this one time wouldn't be an option) seems to be quite a rare thing (I've yet to see a compelling use case), I'm not sure if the effort for this is warranted.
